---
title: "Dimension reduction with Flux Balance Analysis"
author: "Scott Campit"
output: 
  html_notebook:
    toc: yes
    theme: united
---

# Summary
This notebook maps the metabolic fluxes from single-cell Flux Balance Analysis onto the dimension reduction embedding.

# 1. Load Libraries
```{r, results='hide'}
library(tidyverse)
library(umap)
library(Seurat)
library(writexl)
library(ggpubr)
library(gridExtra)
library(extrafont)
loadfonts(device="win")
```

# 2. Load Neutrophil T-SNE embeddings
First, let's load the Neutrophil Seurat object.
```{r, results='hide'}
rm(list=ls())
path_to_neutrophils = "D:/Data/scRNASeq/shea/lung.neut.Robj"
load(path_to_neutrophils, verbose=TRUE)
orig_ident = neut@active.ident
```

## A. Overlay cell subpopulations
Then we'll overlay the clusters onto the current tSNE embedding.
Note: The palette used is colorblind friendly. More information can be found here: https://personal.sron.nl/~pault/

```{r, fig.height=6, fig.width=6, results='hide'}
# Plot based on predicted clusters
neut@active.ident = orig_ident
pA = DimPlot(neut, 
        reduction="tsne",
        dims=c(1, 2),
        cols=c("#4477AA", "#228833", "#AA3377", "#EE6677"),
        label=FALSE,
        repel=TRUE,
        label.size=4,
        pt.size=1.5, 
        na.value = "grey50") + 
        labs(title="Neutrophil subpopulations", 
             x="t-SNE 1", 
             y="t-SNE 2") +
        theme_minimal() +
        theme(plot.title = element_text(size=10),
              legend.text=element_text(size=8),
              axis.title.x=element_text(size=10),
              axis.title.y=element_text(size=10),
              axis.text.x=element_text(size=8),
              axis.text.y=element_text(size=8),
              legend.position=c(0.15, 0.15),
              legend.box="vertical",
              legend.margin=margin(),
              text=element_text(family="sans")) +
        guides(color=guide_legend(override.aes=list(size=2)))

pA
```
## B. Overlay time dimension
Let's also overlay the time dimension onto the T-SNE plot. It's not very clear that we're able to get separation of time points from T-SNE.
```{r, fig.height=6, fig.width=6, results='hide'}
# Plot based on predicted clusters
neut@active.ident = as.factor(neut@meta.data$Time)
levels(neut@active.ident)[1] = "Day 0"
levels(neut@active.ident)[2] = "Day 7"
levels(neut@active.ident)[3] = "Day 14"
levels(neut@active.ident)[4] = "Day 21"
pB = DimPlot(neut, 
        reduction="tsne",
        dims=c(1, 2),
        cols=c("#4477AA", "#228833", "#AA3377", "#EE6677"),
        label=FALSE,
        repel=TRUE,
        label.size=4,
        pt.size=1.5, 
        na.value = "grey50") + 
        labs(title="Neutrophil time-course",
             x="t-SNE 1", 
             y="t-SNE 2") +
        theme_minimal() +
        theme(plot.title = element_text(size=10),
              legend.text=element_text(size=8),
              axis.title.x=element_text(size=10),
              axis.title.y=element_text(size=10),
              axis.text.x=element_text(size=8),
              axis.text.y=element_text(size=8),
              legend.position=c(0.15, 0.15),
              legend.box="vertical",
              legend.margin=margin(),
              text=element_text(family="sans")) +
        guides(color=guide_legend(override.aes=list(size=2)))

pB
```

## C. Overlay metabolic fluxes onto the T-SNE embedding
We'll read in the metabolic flux data computed from single-cell Flux Balance Analysis
```{r, results='hide'}
path_to_neutrophil_flux = 'D:/Analysis/Immune/neutrophil_flux.csv'
flux_data = read_csv(path_to_neutrophil_flux)
flux_data[, 2:ncol(flux_data)] = abs(flux_data[, 2:ncol(flux_data)])
```

Now let's perform some mapping to the original ids
```{r, results='hide'}
# Read in original data
path_to_old_ids = "D:/Data/scRNASeq/shea/neutrophil_colmap.csv"
ids = read_csv(path_to_old_ids)
```

First, we'll grab the day number,
```{r, results='hide'}
# Grab day
ids$day = ids$final_col %>%
    strsplit(" ") %>%
    sapply(tail, 1)
```

Then add a count to the end of that particular time point. Also, reformat it so that it looks like the old mapping.
```{r, results='hide'}
# Iteratively add 1 for all duplicates
ids$day = as.numeric(ids$day)
ids$day = ids$day * 10
ids = data.frame(ids)
ids$new_day = ave(ids$day, ids$final_col, FUN = function(x) x[1] + seq_along(x) - 1)
ids$new_day = ids$new_day - ids$day
ids$day = ids$day / 10
```

Create the final identifier that perfectly imitates the old identifier, and manually set the first occurrences.
```{r, results='hide'}
# Create the new ID
ids$final_day = paste(ids$day, ids$new_day, sep='.')
ids$final_day = as.character(ids$final_day)

# Manually set the first call for a given day
ids$final_day[1] = '0'
ids$final_day[2] = '14'
ids$final_day[19] = '21'
ids$final_day[77] = '7'

# Create the final output
ids$final_col = paste("Neutrophils", ids$final_day, sep="  ")
```

We have to now merge the ids and the metabolic flux data to get a cohesive mapping of cell identifiers to flux.
```{r, results='hide'}
mapped_flux = merge(ids, 
                    flux_data, 
                    by.x='final_col', 
                    by.y='ID')
```

Let's create this dataframe that maps back to the subclusters
```{r, results='hide'}
neutrophil_obj = "D:/Data/scRNASeq/shea/lung.neut.Robj"
load(neutrophil_obj, verbose=TRUE)
tsne_data = data.frame(neut@reductions$tsne@cell.embeddings)
tsne_data$cluster = neut@active.ident
merged_tsne = merge(tsne_data, mapped_flux, by.x=0, by.y='orig_col')
```

Let's intialize some variables we need to declare for the visualization
```{r, results='hide'}
reactions_to_visualize = c("HMR_4418", "HMR_4574", "INSK",
                           "HMR_8486", "HMR_4012", "HMR_4085")
reaction_names = c("Hypoxanthine Phosphoribosyltransferase", 
                   "Inosine Phosphorylase", 
                   "Inosine Kinase",
                   "Adenosine Phosphorylase",
                   "Adenosine Kinase",
                   "Adenine Phosphoribosyltransferase")
colors = c("#CC3311", "#CC3311", "#CC3311",
           "#004488", "#004488", "#004488")
```

Create each T-SNE plot and store them in a list. 
```{r, results='hide'}
p = list()
for(i in 1:length(reactions_to_visualize)){
  p[[i]] = ggplot(merged_tsne, aes(x=tSNE_1, y=tSNE_2, color=merged_tsne[, reactions_to_visualize[i]])) + 
        geom_point(size=1, alpha=0.5) +
        theme_minimal() +
        labs(x="t-SNE 1", 
             y="t-SNE 2",
             title=reaction_names[i], 
             color="Flux") + 
        scale_color_gradient(low="gray", high=colors[i]) +
        theme(plot.title = element_text(size=10),
              legend.text=element_text(size=8),
              axis.title.x=element_text(size=10),
              axis.title.y=element_text(size=10),
              axis.text.x=element_text(size=8),
              axis.text.y=element_text(size=8),
              legend.box="vertical",
              legend.margin=margin())
}
```

Create and visualize the final T-SNE flux plot object.
```{r, fig.width=12, fig.height=6, results='hide'}
pC = scater::multiplot(p[[1]], p[[6]], p[[2]], p[[4]], p[[3]], p[[5]], cols=3)
pC
```

Finally, let's assemble the final figure:
```{r, results='hide'}
pX = ggarrange(pA, pB, 
               ncol=1, nrow=2, 
               labels=c("A", "B"))
pY = ggarrange(pC, 
               ncol=1, 
               labels="C",
               common.legend=FALSE)
ggarrange(pX, pY, 
          nrow=1, ncol=2, 
          widths=c(6, 24), heights=c(12, 12),
          common.legend=FALSE)
```

# 3. Get Metabolic Reactions That Are Significantly Upregulated in T-SNE Embedding
To overlay the metabolic fluxes, we'll perform differential metabolic flux analysis using Seurat.

## A. Format FBA Data
First, we need to format the data, so that it matches the DGE Matrix, where rows are reactions and columns are cell IDs.
```{r, results='hide'}
fba = mapped_flux

# Remove unnecessary columns
fba[, c("final_col", "day", "new_day", "final_day")] = NULL

# Map to original IDs
rownames(fba) = fba$orig_col
fba$orig_col = NULL

# Reformat
fba = t(fba)
fba = as.sparse(fba)
```

Now we need to make sure the expression matrix has the same cells as the fba matrix, and also ensure the `active.ident` and the `Time` datapoints we'll map to.
```{r, results='hide'}
# Get names and indices
intersect_cells = intersect(colnames(neut@assays$RNA@counts), colnames(fba))
cell_idx = which(colnames(neut@assays$RNA@counts) %in% intersect_cells)

# Reformat data
neut = subset(neut, cells=intersect_cells)
```

## B. Create FBA Assay in Seurat Object
This incorporates FBA into the Seurat Object.
```{r, results='hide'}
fba = fba[, intersect_cells]
neut[["FBA"]] = CreateAssayObject(counts=fba)
```

## C. Perform data preprocessing for metabolic fluxes
We'll now perform some data preprocessing steps for the metabolic fluxes. Let's define the function with the following preprocessing steps:
  1. Log normalization
  2. Data scaling

```{r, results='hide'}
preprocess_flux = function(gene_obj, assay){
  # Log normalize data
  gene_obj = NormalizeData(gene_obj, 
                           assay=assay,
                           normalization.method='LogNormalize', 
                           scale.factor=10000)
  
  # Scale data
  gene_obj = ScaleData(gene_obj, 
                       assay=assay)
  return(gene_obj)
}
```

Now let's perform the data preprocessing:
```{r, results='hide'}
neut = preprocess_flux(neut, assay="FBA")
```

## D. Get Metabolic Markers
We'll get metabolic markers based on subpopulations.
```{r, results='hide'}
#neut@active.ident = as.factor(neut@meta.data$Time)
subpop_reactions = FindAllMarkers(neut,
                                  assay="FBA",
                                  only.pos=TRUE,
                                  min.pct=0.25,
                                  logfc.threshold=0.25,
                                  test.use="MAST")
#write.table(subpop_reactions, "clipboard", sep="\t")
```

## E. Plot heatmap of differentially active metabolic reactions from T-SNE Neutrophil Subpopulations
These are the reaction names we need to feed into the visualization.
```{r, results='hide'}
reaction_names = c("Adenosine Phosphorylase", # HMR_8486
                   "Adenosine Kinase", # HMR_4012
                   "Adenine Phosphoribosyltransferase", # HMR_4085
                   "Glutamate Dehydrogenase", # HMR_3802
                   "Inosine Phosphorylase", # HMR_4574
                   "Inosine Kinase", # (INSK)
                   "Hypoxanthine Phosphoribosyltransferase", # HMR_4418
                   "Isocitrate Dehydrogenase (NADPH)",# HMR_4586
                   "Isocitrate Dehydrogenase (NADH)" # HMR_4588
                   )
```

 it onto the HeatMap
```{r, results='hide'}
DoHeatmap(object=neut,
          assay="FBA",
          slot='data',
          label=FALSE,
          draw.lines=TRUE,
          lines.width=10,
          features=subpop_reactions$gene,
          group.colors=c("#4477AA", "#228833", "#AA3377", "#EE6677")) +
          scale_fill_gradient(name="Metabolic flux",
                              low='gray', high='#CC3311') +
          scale_y_discrete(labels=c("HMR-8486"="Adenosine Phosphrylase", 
                                    "HMR-4012"="Adenosine Kinase",
                                    "HMR-4085"="Adenine Phosphoribosyltransferase",
                                    "HMR-3802"="Glutamate Dehydrogenase",
                                    "HMR-4574"="Inosine Phosphorylase",
                                    "INSK"="Inosine Kinase",
                                    "HMR-4418"="Hypoxanthine Phosphoribosyltransferase",
                                    "HMR-4586"="Isocitrate Dehydrogenase (NADPH)",
                                    "HMR-4588"="Isocitrate Dehydrogenase (NADH)")) +
          theme(text=element_text(family='sans', size=16),
                legend.text=element_text(size=16),
                axis.title.x=element_text(size=20),
                axis.title.y=element_text(size=20),
                axis.text.x=element_text(size=16),
                axis.text.y=element_text(size=16),
                legend.title=element_text(size=20)) +
          labs(colour="Neutrophil subtypes")
```

## F. Get gene markers for each subpopulation
This will be used to see if there is correspondence in the UMAP dataset
```{r, results='hide'}
subpop_genes_tsne = FindAllMarkers(neut,
                              assay="RNA",
                              only.pos=TRUE,
                              min.pct=0.25,
                              logfc.threshold=0.25,
                              test.use="MAST")
```

# 4. UMAP Analysis 

## A. Load Dataset
Load original lung object.
```{r, results='hide'}
all_lung_data = "D:/Data/scRNASeq/shea/lung.Robj"
load(all_lung_data, verbose=TRUE)
```

We'll now subset the neutrophil data
```{r, results='hide'}
orig.indent = lung@active.ident
celltype.list = SplitObject(lung, split.by="CellType")

# Create separate Seurat objects to visualize.
neutrophils   = celltype.list$Neutrophils
```

This formats the FBA data.
```{r, results='hide'}
fba = mapped_flux

# Remove unnecessary columns
fba[, c("final_col", "day", "new_day", "final_day")] = NULL

# Map to original IDs
rownames(fba) = fba$orig_col
fba$orig_col = NULL

# Reformat
fba = t(fba)
fba = as.sparse(fba)
```

Now we need to make sure the expression matrix has the same cells as the fba matrix, and also ensure the `active.ident` and the `Time` datapoints we'll map to.
```{r, results='hide'}
# Get names and indices
intersect_cells = intersect(colnames(neutrophils@assays$RNA@counts), colnames(fba))
cell_idx = which(colnames(neutrophils@assays$RNA@counts) %in% intersect_cells)

# Reformat data
neutrophils = subset(neutrophils, cells=intersect_cells)
```

## B. Create FBA Assay in Seurat Object
This incorporates FBA into the Seurat Object.
```{r, results='hide'}
fba = fba[, intersect_cells]
neutrophils[["FBA"]] = CreateAssayObject(counts=fba)
```

Then, we create the metabolic flux object. Run if you haven't already in the T-SNE section.
```{r, results='hide'}
# Create identity
#mapped_flux = mapped_flux[, -c(final_col, day, new_day, final_day)]
# Cells need to be cols and rows need to be reactions
#flux_data = t(flux_data)

# Set column names
#names(flux_data) = lapply(flux_data[1, ], as.character)
#colnames(flux_data) = as.character(flux_data["ID", ])
#flux_data = flux_data[-1, ] 
```

## C. Preprocess Neutrophil Dataset
This function will find subpopulations within a given cell type. This can be used later.

```{r, results='hide'}
find_subpopulations = function(gene_obj){
  
  # Identify variable features
  gene_obj = FindVariableFeatures(object=gene_obj, 
                              mean.function=ExpMean,
                              dispersion.function=LogVMR,
                              selection.method='vst',
                              x.low.cutoff=-Inf,
                              x.high.cutoff=Inf,
                              y.cutoff=Inf,
                              nfeatures = 2000)
  
  # Log normalize data
  gene_obj = NormalizeData(gene_obj, 
                       normalization.method='LogNormalize', 
                       scale.factor=10000)
  
  # Scale data
  all.genes = rownames(gene_obj)
  gene_obj = ScaleData(gene_obj, 
                   features=all.genes)
  
  gene_obj = SCTransform(gene_obj)

  
  # Run PCA
  gene_obj = RunPCA(object=gene_obj,
                npcs=50,
                verbose=FALSE)
  
  # Find shortest nearest neighbors and clusters
  gene_obj = FindNeighbors(gene_obj, 
                       dims=1:50,
                       reduction='pca')
  gene_obj = FindClusters(gene_obj, 
                      resolution=0.5)
  
  # Perform T-SNE
  gene_obj = RunTSNE(gene_obj,
                     dims=1:50)
  
  # Perform UMAP
  gene_obj = RunUMAP(gene_obj, 
                 dims=1:50)
  
  return(gene_obj)
}
```

Determine the subpopulations with the neutrophils
```{r, results='hide', include=FALSE}
neutrophils = find_subpopulations(neutrophils)
orig_ident = neutrophils@active.ident
```

## D. Plot the UMAP embedding for subpopulations.
```{r, results='hide'}
# Plot based on predicted clusters
neutrophils@active.ident = orig_ident
pA = DimPlot(neutrophils, 
        reduction="umap",
        dims=c(1, 2),
        cols=c("#4477AA", "#228833", "#AA3377", "#EE6677"),
        label=FALSE,
        repel=TRUE,
        label.size=4,
        pt.size=1.5, 
        na.value = "grey50") + 
        labs(title="Neutrophil subpopulations", 
             x="UMAP 1", 
             y="UMAP 2") +
        theme_minimal() +
        theme(plot.title = element_text(size=10),
              legend.text=element_text(size=8),
              axis.title.x=element_text(size=10),
              axis.title.y=element_text(size=10),
              axis.text.x=element_text(size=8),
              axis.text.y=element_text(size=8),
              legend.position=c(0.15, 0.85),
              legend.box="vertical",
              legend.margin=margin(),
              text=element_text(family="sans")) +
        guides(color=guide_legend(override.aes=list(size=2)))

pA
```

## E. Plot the UMAP embedding for time-course
```{r, results='hide'}
# Plot based on predicted clusters
neutrophils@active.ident = as.factor(neutrophils@meta.data$Time)
levels(neutrophils@active.ident)[1] = "Day 0"
levels(neutrophils@active.ident)[2] = "Day 7"
levels(neutrophils@active.ident)[3] = "Day 14"
levels(neutrophils@active.ident)[4] = "Day 21"
pB = DimPlot(neutrophils, 
        reduction="umap",
        dims=c(1, 2),
        cols=c("#4477AA", "#228833", "#AA3377", "#EE6677"),
        label=FALSE,
        repel=TRUE,
        label.size=4,
        pt.size=1.5, 
        na.value = "grey50") + 
        labs(title="Neutrophil time-course", 
             x="UMAP 1", 
             y="UMAP 2") +
        theme_minimal() +
        theme(plot.title = element_text(size=10),
              legend.text=element_text(size=8),
              axis.title.x=element_text(size=10),
              axis.title.y=element_text(size=10),
              axis.text.x=element_text(size=8),
              axis.text.y=element_text(size=8),
              legend.position=c(0.15, 0.85),
              legend.box="vertical",
              legend.margin=margin(),
              text=element_text(family="sans")) +
        guides(color=guide_legend(override.aes=list(size=2)))

pB
```
## F. Overlay metabolic fluxes onto UMAP embedding
Now that we know everything is working, let's grab the data.
```{r, results='hide'}
neutrophils@active.ident = orig_ident
umap_data = data.frame(neutrophils@reductions$umap@cell.embeddings)
umap_data$cluster = neutrophils@active.ident
merged_umap = merge(umap_data, mapped_flux, by.x=0, by.y='orig_col')
```


```{r, results='hide'}
q1 = ggplot(merged_umap, aes(x=UMAP_1, y=UMAP_2, color=HMR_4418))+ 
        geom_point(size=1, alpha=0.5) +
        theme_minimal() +
        labs(x="UMAP 1", y="UMAP 2",
             title="Hypoxanthine Phosphoribosyltransferase", color="Flux")  + 
        scale_color_gradient(low="gray", high="#CC3311") +
        theme(plot.title = element_text(size=10),
              legend.text=element_text(size=8),
              axis.title.x=element_text(size=10),
              axis.title.y=element_text(size=10),
              axis.text.x=element_text(size=8),
              axis.text.y=element_text(size=8),
              legend.box="vertical",
              legend.margin=margin())

q2 = ggplot(merged_umap, aes(x=UMAP_1, y=UMAP_2, color=HMR_4574))+ 
        geom_point(size=1, alpha=0.5) +
        theme_minimal() +
        labs(x="UMAP 1", y="UMAP 2",
             title="Inosine Phosphorylase", color="Flux") + 
        scale_color_gradient(low="gray", high="#CC3311") +
        theme(plot.title = element_text(size=10),
              legend.text=element_text(size=8),
              axis.title.x=element_text(size=10),
              axis.title.y=element_text(size=10),
              axis.text.x=element_text(size=8),
              axis.text.y=element_text(size=8),
              legend.box="vertical",
              legend.margin=margin())

q3 = ggplot(merged_umap, aes(x=UMAP_1, y=UMAP_2, color=INSK))+ 
        geom_point(size=1, alpha=0.5) +
        theme_minimal() +
        labs(x="UMAP 1", y="UMAP 2",
             title="Inosine Kinase", color="Flux") + 
        scale_color_gradient(low="gray", high="#CC3311") +
        theme(plot.title = element_text(size=10),
              legend.text=element_text(size=8),
              axis.title.x=element_text(size=10),
              axis.title.y=element_text(size=10),
              axis.text.x=element_text(size=8),
              axis.text.y=element_text(size=8),
              legend.box="vertical",
              legend.margin=margin())

p2 = ggplot(merged_umap, aes(x=UMAP_1, y=UMAP_2, color=HMR_8486))+ 
        geom_point(size=1, alpha=0.5) +
        theme_minimal() +
        labs(x="UMAP 1", y="UMAP 2",
             title="Adenosine Phosphorylase", color="Flux") + 
        scale_color_gradient(low="gray", high="#004488") +
        theme(plot.title = element_text(size=10),
              legend.text=element_text(size=8),
              axis.title.x=element_text(size=10),
              axis.title.y=element_text(size=10),
              axis.text.x=element_text(size=8),
              axis.text.y=element_text(size=8),
              legend.box="vertical",
              legend.margin=margin())

p3 = ggplot(merged_umap, aes(x=UMAP_1, y=UMAP_2, color=HMR_4012))+ 
        geom_point(size=1, alpha=0.5) +
        theme_minimal() +
        labs(x="UMAP 1", y="UMAP 2",
             title="Adenosine Kinase", color="Flux") + 
        scale_color_gradient(low="gray", high="#004488") +
        theme(plot.title = element_text(size=10),
              legend.text=element_text(size=8),
              axis.title.x=element_text(size=10),
              axis.title.y=element_text(size=10),
              axis.text.x=element_text(size=8),
              axis.text.y=element_text(size=8),
              legend.box="vertical",
              legend.margin=margin())

p4 = ggplot(merged_umap, aes(x=UMAP_1, y=UMAP_2, color=HMR_4085))+ 
        geom_point(size=1, alpha=0.5) +
        theme_minimal() +
        labs(x="UMAP 1", y="UMAP 2",
             title="Adenine Phosphoribosyltransferase", color="Flux") + 
        scale_color_gradient(low="gray", high="#004488") +
        theme(plot.title = element_text(size=10),
              legend.text=element_text(size=8),
              axis.title.x=element_text(size=10),
              axis.title.y=element_text(size=10),
              axis.text.x=element_text(size=8),
              axis.text.y=element_text(size=8),
              legend.box="vertical",
              legend.margin=margin())
```

Create the plot object.
```{r, fig.width=12, fig.height=6, results='hide'}
pC = scater::multiplot(q1, p4, q2, p2, q3, p3, cols=3)
pC
```

Finally, let's assemble the final figure:
```{r, results='hide'}

pX = ggarrange(pA, pB, 
               ncol=1, nrow=2, 
               labels=c("A", "B"))
pY = ggarrange(pC, 
               ncol=1, 
               labels="C",
               common.legend=FALSE)
ggarrange(pX, pY, 
          nrow=1, ncol=2, 
          widths=c(6, 24), heights=c(12, 12),
          common.legend=FALSE)
```

# 5. Get metabolic reactions that are differentially expressed

## A. Perform Data Preprocessing
Now let's perform the data preprocessing:
```{r, results='hide'}
neutrophils = preprocess_flux(neutrophils, assay="FBA")
```

## B. Get Metabolic Markers
We'll get metabolic markers based on subpopulations.
```{r, results='hide'}
#neutrophils@active.ident = orig_ident
neutrophils@active.ident = as.factor(neutrophils@meta.data$Time)
subpop_reactions = FindAllMarkers(neutrophils,
                                  assay="FBA",
                                  test.use="MAST")
tmp = as.data.frame(table(subpop_reactions$gene))
tmp = tmp[tmp$Freq < 2, ]

subpop_reactions = merge(subpop_reactions, tmp, by.x='gene', by.y='Var1')
#write.table(subpop_reactions, "clipboard", sep="\t")
```

## C. Plot heatmap of differentially active metabolic reactions from UMAP Neutrophil Subpopulations
Plot it onto the HeatMap
```{r, results='hide'}
reaction_names = c("Adenosine Phosphorylase", # HMR_8486
                   "Adenosine Kinase", # HMR_4012
                   "Adenine Phosphoribosyltransferase", # HMR_4085
                   "Glutamate Dehydrogenase", # HMR_3802
                   "Inosine Phosphorylase", # HMR_4574
                   "Inosine Kinase", # (INSK)
                   "Hypoxanthine Phosphoribosyltransferase", # HMR_4418
                   "Isocitrate Dehydrogenase (NADPH)",# HMR_4586
                   "Isocitrate Dehydrogenase (NADH)" # HMR_4588
                   )

DoHeatmap(object=neutrophils,
          assay="FBA",
          slot='data',
          label=FALSE,
          draw.lines=TRUE,
          lines.width=10,
          features=subpop_reactions$gene,
          group.colors=c("#4477AA", "#228833", "#AA3377", "#EE6677")) +
          scale_fill_gradient(name="Metabolic flux",
                              low='gray', high='#CC3311') +
          scale_y_discrete(labels=c("HMR-8486"="Adenosine Phosphrylase", 
                                    "HMR-4012"="Adenosine Kinase",
                                    "HMR-4085"="Adenine Phosphoribosyltransferase",
                                    "HMR-3802"="Glutamate Dehydrogenase",
                                    "HMR-4574"="Inosine Phosphorylase",
                                    "INSK"="Inosine Kinase",
                                    "HMR-4418"="Hypoxanthine Phosphoribosyltransferase",
                                    "HMR-4586"="Isocitrate Dehydrogenase (NADPH)",
                                    "HMR-4588"="Isocitrate Dehydrogenase (NADH)")) +
          theme(text=element_text(family='sans', size=16),
                legend.text=element_text(size=16),
                axis.title.x=element_text(size=20),
                axis.title.y=element_text(size=20),
                axis.text.x=element_text(size=16),
                axis.text.y=element_text(size=16),
                legend.title=element_text(size=20)) +
          labs(colour="Neutrophil subtypes")
```

## D. Get gene markers for each subpopulation
This will be used to see if there is correspondance in the UMAP dataset
```{r, results='hide'}
subpop_genes_umap = FindAllMarkers(neutrophils,
                              assay="RNA",
                              only.pos=TRUE,
                              min.pct=0.25,
                              logfc.threshold=0.25,
                              test.use="MAST")
```
# 6. Merge the T-SNE and UMAP embedding together to see if there is overlap.
I need to get unique markers for each cluster. This is how I'll do it.
```{r, results='hide'}
# Remove all genes that are duplicated in UMAP
dup_umap = subpop_genes_umap$gene[duplicated(subpop_genes_umap$gene)]
subpop_genes_umap = subpop_genes_umap[which(rownames(subpop_genes_umap) %in% dup_umap), ]

# Remove all genes that are duplicated in T-SNE 
dup_tsne = subpop_genes_tsne$gene[duplicated(subpop_genes_tsne$gene)]
subpop_genes_tsne = subpop_genes_tsne[which(rownames(subpop_genes_tsne) %in% dup_tsne), ]

# Get intersection
merged_clusters = merge(subpop_genes_tsne, subpop_genes_umap, by='gene')
```